### DISCUSSION TITLE
SSR on version 12.1.6

### DISCUSSION THREAD

**Discussion by anhthien-bic** [vercel/next.js#37818] [Help]
UPDATE: I am going to create new issue or feature request

**Comment by icyJoseph** [comment]
Could you make a small reproduction repo? Or share a link to a repo with this issue?

**Reply by icyJoseph** [reply]
I don't really like this architecture, but I think it works out fine if you do something like:

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";
import { useCallback, useEffect, useState } from "react";
import LoginPage from "./login";

const isClient = () =>
  typeof window !== "undefined" && typeof document !== "undefined";

const Pending = "pending";
const LoggedIn = "loggedin";
const LoggedOut = "loggedout";

type UserState = typeof Pending | typeof LoggedIn | typeof LoggedOut;

function MyApp({ Component, pageProps }: AppProps) {
  const [loginState, setLoginState] = useState<UserState>(Pending);

  const checkLogin = useCallback(() => {
    setTimeout(() => setLoginState(LoggedIn), 1000);
  }, []);

  useEffect(() => {
    if (loginState === LoggedIn) return;

    checkLogin();
  }, [checkLogin, loginState]);

  // @ts-ignore
  const getLayout = Component.getLayout ?? ((page) => page);

  if (isClient()) {
    if (loginState === Pending) return getLayout(<Component {...pageProps} />);

    return loginState === LoggedIn ? (
      getLayout(<Component {...pageProps} />)
    ) : (
      <LoginPage />
    );
  }

  return getLayout(<Component {...pageProps} />);
}

export default MyApp;
```

I'd say it might be better to use rewrites with middleware, but that depends on how we arrive to the conclusion that the user is logged in. 

Let me know if the above works out.

**Reply by icyJoseph** [reply]
I'll drop an example later today :)

**Reply by icyJoseph** [reply]
> I did use your code and get the error Hydration fail as expected 😂

Strange, I don't get that error at all, not even in production mode. Could you share what's broken in yours, with my suggestion?

> The old version React17 works fine but only this case broken on React18.

I think that's not really true, the error was just silent before React 18....

**Reply by icyJoseph** [reply]
### With middleware

This one assumes that you are using a cookie to label the user session.

```ts
// pages/_middleware.ts
import { type NextRequest, NextResponse } from "next/server";

export const middleware = async (req: NextRequest) => {
  const { pathname } = req.nextUrl;

  if (
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") // exclude all files in the public folder
  )
    return NextResponse.next();

  if (pathname === "/") {
    const isBot = req.ua?.isBot ?? false;

    if (isBot) {
      // for bots, let the request go through
      return NextResponse.next();
    }

    // verify session
    const cookie = req.cookies["foo"];

    // here you'd do some verify on the cookie, or even connect
    // to another service to confirm the session validity

    if (!cookie) {
      // show the login page in place of `/` (url still shows '/')
      req.nextUrl.pathname = "/login";
      return NextResponse.rewrite(req.nextUrl);
    }

    return NextResponse.next();
  }
};
```

Now, you don't need any code, other than the per-page layout getter, in `_app`:

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";

function MyApp({ Component, pageProps }: AppProps) {
  // @ts-ignore
  const getLayout = Component.getLayout ?? ((page) => page);

  return getLayout(<Component {...pageProps} />);
}

export default MyApp;
```

When I do a request:

```http
GET http://localhost:3000
Cookie: foo=tomato
```

I see the page, but if I omit the cookie

```http
GET http://localhost:3000
```

I get the HTML for the login page, still at the `/` route. I can also just navigate to the `/login` page naturally.

And If I was a bot, in this list: `/Googlebot|Mediapartners-Google|AdsBot-Google|googleweblight|Storebot-Google|Google-PageRenderer|Bingbot|BingPreview|Slurp|DuckDuckBot|baiduspider|yandex|sogou|LinkedInBot|bitlybot|tumblr|vkShare|quora link preview|facebookexternalhit|facebookcatalog|Twitterbot|applebot|redditbot|Slackbot|Discordbot|WhatsApp|SkypeUriPreview|ia_archiver/i`, then I'd just see the page at `/`.


To prevent people from trying to game this, by setting a custom user agent, you can still have a `useEffect`, in `_app`. Most importantly though, you should verify bot IP's inside middleware as well, which is [pretty much recommended](https://developers.google.com/search/docs/advanced/crawling/verifying-googlebot) by every vendor that has a bot/crawler.

```tsx
function MyApp({ Component, pageProps }: AppProps) {
  useEffect(() => {
    const controller = new AbortController();

    fetch("/api/verify-user", { signal: controller.signal })
      .then((res) => {
        if (!res.ok) return Router.replace("/login");
      })
      .catch((_err) => {
        if (controller.signal.aborted) {
          console.log("Cancelled user verification");
        } else {
          console.error("Error while verify user");
          Router.replace("/login");
        }
      });

    return () => {
      controller.abort();
    };
  }, []);

  // @ts-ignore
  const getLayout = Component.getLayout ?? ((page) => page);

  return getLayout(<Component {...pageProps} />);
}
```

You could even run the check on every `routeChangeStart` events from the `next/router`.

I think this is slightly better than juggling around with different HTML outputs based on user state. Instead, there's a clear division of responsibilities between each page, the middleware, and the `_app` that controls things at runtime. React renders UI, Next.js pre-renders all, and middleware takes care of routing users.

**Reply by anhthien-bic** [reply]
Yeah, I will try your code, thank you so much

**Reply by icyJoseph** [reply]
You want to show the login page to bots too, is that correct?

**Reply by anhthien-bic** [reply]
yeah :D

**Reply by icyJoseph** [reply]
I think middleware can take care of all these cases. Even for bots, you'd just rewrite to `login` page, by modifying what the example I gave does. Rewrites keep the `url` in the address bar, but swap the content that's is shown to the user/bot.

**Reply by icyJoseph** [reply]
Hi, could you paste a snippet of your middleware? or share a repo?

I wonder if your page, for `/` is getting pre-rendered with the contents for login. 

When you build your application, does it say that it pre-rendered your `/` page?

**Reply by icyJoseph** [reply]
Mmm going through what was said before, and the initial architecture, I am still not sure about this.

Could you update the repository you had shared? Also, what happens if you just put `App.getInitialProps = () => {return {}}` (**EDIT**: I forgot the return here) in `_app`? Replace `App` with the name of your custom App. I expect this to disable static generation of pages that do not implement `getStaticProps` or `getServerSideProps`. So the `/` page would be generated on demand.

**Reply by anhthien-bic** [reply]
ok, I will update the repo tonight, pls wait 👍

**Reply by icyJoseph** [reply]
Mmm I see, so in this case your login function is just getting something from a network call, and uses that result to decide if we are logged in, but it doesn't really redirect or anything like that.

I had expected that on a submit handler, you'd redirect users back `/` and that should've done the trick. 

I've been going over this once again, perhaps **we have to start from zero**. 

We have this list of requirements:

```markdown
1. I have a home page at "/"
2. If request to route "/" is SSR only (check window is undefined and I dont use getSSProps, getStaticProps,... in this page, I use component only) > render component Login Page (1)
3. If request to route "/" is CSR and user logged in > render Page "/"
4. Else if request to route "/" is CSR and user NOT logged in > render component Login Page (1)
```

The first thing to understand is that, the hydration error, is legit, and it is not something we can do much about. 

The HTML rendered by the server, must match the first frame rendered by the client. **We cannot design an application that goes against this.** 

When you say `If request to route "/" is SSR only` I think I've been making assumptions, but what do you actually mean by this? Let's for example think, if I am a visitor to your site, with JavaScript enabled, I go to `/`, either I get HTML generated at build time, or there's HTML generated on demand. As said before, in both cases, the HTML sent by the server, and the first frame generated on the client must match.

So this means that as soon as the first client side frame is rendered, also called hydration, we can run `useEffect` and other logic, which could could change the HTML sent by the server to match some kind of client side logic, for example, whether or not you are logged in.

I could also be a crawler/bot that doesn't run JavaScript. In this case I go to `/`, and I get the HTML generated at build time, or HTML generated on demand, but I'll never run any JavaScript, so I parse what I received, and there's no JavaScript that kicks in to ever change this.

Note that for HTML generated on demand, on both cases, we could check if the user is logged in, with cookies for example, and respond with proper HTML for that logged in user.

With this considerations, what do you actually mean in point 2?

`3. If request to route "/" is CSR and user logged in > render Page "/"` and `4. Else if request to route "/" is CSR and user NOT logged in > render component Login Page` These make me think of a client side transition, where in `_app`, inside `useEffect` we could see that if the current route is `/`, and if there's no user logged in, `replace` it with `/login`, for example.

From looking at the conversation I think 3, and 4 have been pretty clear, but the issue comes from what's point 2 actually defining. Also, how do you decide when a user is logged in?

**Reply by icyJoseph** [reply]
Let's see, I've played with this a bit, let me know how this works out for you:

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";
import LoginPage from "./login";
import { useRouter } from "next/router";

const Pending = "pending";
const LoggedIn = "loggedin";
const LoggedOut = "loggedout";

type UserState = typeof Pending | typeof LoggedIn | typeof LoggedOut;

function MyApp({ Component, pageProps }: AppProps) {
  // @ts-ignore
  const getLayout = Component.getLayout ?? ((page) => page);
  const router = useRouter();
  console.log("router", router);

  if (router.query.loggedIn === "required") return <LoginPage />;

  return getLayout(<Component {...pageProps} />);
}

MyApp.getInitialProps = () => {
  return {};
};

export default MyApp;
```

The `getInitialProps` is still needed, and `/` shouldn't implement `getStaticProps`, because otherwise, it is pre-rendered with empty query in `router.query.loggedIn`.

Then change the middleware to:

```ts
import { type NextRequest, NextResponse } from "next/server";

export default async function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  if (
    pathname.includes(".") || // exclude all files in the public folder
    pathname.includes("/_next") || // exclude all next internal routes
    pathname.startsWith("/static") || // exclude static files
    pathname.startsWith("/api") //  exclude all API routes
  ) {
    return NextResponse.next();
  }

  // check is user logged in
  const isLoggedIn = req.cookies.get("logged_in");

  if (!isLoggedIn) {
    req.nextUrl.searchParams.set("loggedIn", "required");
    req.nextUrl.pathname = "/";
    return NextResponse.rewrite(req.nextUrl); // rewrite login page
  }

  // user is logged in
  if (pathname === "/login") {
    // user is logged in and path is login > redirect to home page
    req.nextUrl.pathname = "/";
    return NextResponse.redirect(req.nextUrl);
  }

  return NextResponse.next();
}
```

What do you think?
